projects/media-viewer/src/lib/toolbar/side-bar/outline-view/outline-view.component.ts

```diff
@@ -0,0 +1,139 @@
+import {
+  AfterContentInit, Component, ElementRef, Input, OnChanges, SimpleChanges, ViewChild,
+  ViewEncapsulation
+} from '@angular/core';
+import { PdfJsWrapperFactory } from '../../../viewers/pdf-viewer/pdf-js/pdf-js-wrapper.provider';
+
+@Component({
+  selector: 'mv-outline-view',
+  templateUrl: './outline-view.component.html',
+  styleUrls: ['../../../styles/main.scss'],
+  encapsulation: ViewEncapsulation.None
+})
+export class OutlineViewComponent implements OnChanges, AfterContentInit {
+
+  @Input() url: string;
+  @ViewChild('outlineContainer') outlineContainer: ElementRef;
+
+  private pdfDocumentOutline: pdfjsOutlineViewer.PDFOutlineViewer;
+  private outline: {};
+
+  constructor(private readonly pdfJsWrapperFactory: PdfJsWrapperFactory) {
+  }
+
+  async ngAfterContentInit() {
+    this.renderOutlineView();
+  }
+
+  async ngOnChanges(changes: SimpleChanges) {
+    if (changes.url) {
+      this.outline = {};
+      await this.renderOutlineView();
+    }
+  }
+
+  async renderOutlineView() {
+    this.pdfDocumentOutline = this.pdfJsWrapperFactory.createDocumentOutline(this.outlineContainer);
+    const pdf = await pdfjsLib.getDocument({
+      url: this.url,
+      cMapUrl: 'assets/minified/cmaps',
+      cMapPacked: true,
+      withCredentials: true
+    });
+
+    this.outline = await pdf.getOutline();
+    this.render({ outline: this.outline });
+  }
+
+  bindOutlineLink(element, { url, newWindow, dest }) {
+    const linkService = this.pdfDocumentOutline.linkService;
+
+    element.href = linkService.getDestinationHash(dest);
+    element.onclick = () => {
+      if (dest) {
+        linkService.navigateTo(dest);
+      }
+      return false;
+    };
+  }
+
+  setStyles(element, { bold, italic }) {
+    let styleStr = '';
+    if (bold) {
+      styleStr += 'font-weight: bold;';
+    }
+    if (italic) {
+      styleStr += 'font-style: italic;';
+    }
+    if (styleStr) {
+      element.setAttribute('style', styleStr);
+    }
+  }
+
+  addToggleButton(div, { count, items, }) {
+    const toggler = document.createElement('div');
+    toggler.className = 'outlineItemToggler outlineItemsHidden';
+
+    toggler.onclick = (evt) => {
+      evt.stopPropagation();
+      toggler.classList.toggle('outlineItemsHidden');
+    };
+
+    div.insertBefore(toggler, div.firstChild);
+  }
+
+  resetOutline() {
+    let firstNode = this.outlineContainer.nativeElement.firstElementChild;
+    while (firstNode) {
+      this.outlineContainer.nativeElement.removeChild(firstNode);
+      firstNode = this.outlineContainer.nativeElement.firstElementChild;
+    }
+  }
+
+  render({ outline, }) {
+    let outlineCount = 0;
+
+    if (this.outline) {
+      this.resetOutline();
+    }
+    this.outline = outline || null;
+
+    const fragment = document.createDocumentFragment();
+    const queue = [{ parent: fragment, items: this.outline, }];
+    let hasAnyNesting = false;
+    while (queue.length > 0) {
```

Comment:
Okay, so what we have here is a tree structure rather than a queue.

The best way to iterate a tree is through recursion, so I would split this loop out into it's own method that took an arbitrary point in the tree, created the dom element and then scanned for children. Everything above this line would stay as a bootstrap for the recursive function. For example:

```
interface TreeNode {
  title: string | null, // this seems to be null and default to outline + outlineCount + 1
  parent: TreeNode //or you could use Fragment if that's preferred
  children: TreeNode[]
}

function renderTree(node: TreeNode) {
  this.createTreeNodeDomElement(node.title);

  for (const child of node.children) {
    this.renderTree(child);
  }
}
```

Obviously this a vast simplification of what is happening but it should suffice to give you the idea. The outlineCount can be passed to the `renderTree` method if it is required.